src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/CalculateContamination.java

```diff
@@ -89,73 +108,144 @@
     private final double lowCoverageRatioThreshold = DEFAULT_LOW_COVERAGE_RATIO_THRESHOLD;
 
     @Argument(fullName= "highCoverageRatioThreshold",
-            doc="The maximum coverage relative to the median.", optional = true)
+            doc="The maximum coverage relative to the mean.", optional = true)
     private final double highCoverageRatioThreshold = DEFAULT_HIGH_COVERAGE_RATIO_THRESHOLD;
-    
+
+    private static final double SEGMENTATION_KERNEL_VARIANCE = 0.025;
+
     private static final BiFunction<PileupSummary, PileupSummary, Double> SEGMENTATION_KERNEL = (ps1, ps2) -> {
         final double maf1 = FastMath.min(ps1.getAltFraction(), 1 - ps1.getAltFraction());
         final double maf2 = FastMath.min(ps2.getAltFraction(), 1 - ps2.getAltFraction());
-        return FastMath.exp(-MathUtils.square(maf1 - maf2));
+        return FastMath.exp(-MathUtils.square(maf1 - maf2)/(2 * SEGMENTATION_KERNEL_VARIANCE));
     };
 
     @Override
     public Object doWork() {
-        final List<PileupSummary> pileupSummaries = filterSites(PileupSummary.readPileupSummaries(inputPileupSummariesTable));
-        final List<PileupSummary> matchedPileupSummaries = matchedPileupSummariesTable == null ? null
-                : filterSites(PileupSummary.readPileupSummaries(matchedPileupSummariesTable));
+        final List<PileupSummary> sites = filterSites(PileupSummary.readFromFile(inputPileupSummariesTable));
 
         // used the matched normal to genotype if available
-        final List<PileupSummary> pileupSummariesForGenotyping = matchedPileupSummaries == null ? pileupSummaries : matchedPileupSummaries;
-        final List<PileupSummary> homAltSites = findHomAltSites(pileupSummariesForGenotyping);
+        final List<PileupSummary> genotypingSites = matchedPileupSummariesTable == null ? sites :
+                filterSites(PileupSummary.readFromFile(matchedPileupSummariesTable));
+
+        final List<List<PileupSummary>> genotypingSegments = findSegments(genotypingSites);
 
-        final List<PileupSummary> homAltPileups = matchedPileupSummaries == null ? homAltSites : getCorrespondingHomAltSites(homAltSites, pileupSummaries);
 
-        final Pair<Double, Double> contaminationAndError = calculateContamination(homAltPileups);
-        ContaminationRecord.writeContaminationTable(Arrays.asList(new ContaminationRecord(ContaminationRecord.Level.WHOLE_BAM.toString(), contaminationAndError.getLeft(), contaminationAndError.getRight())), outputTable);
+        List<PileupSummary> homAltGenotypingSites = new ArrayList<>();
+        final MutableDouble genotypingContamination = new MutableDouble(INITIAL_CONTAMINATION_GUESS);
+
+        for (int iteration = 0; iteration < MAX_ITERATIONS; iteration++) {
```

Comment:
There is no code already for handling max iterations and a convergence criteria?  If not, or this would take too long to integrate then disregard this comment.

Reply:
There is no code